### PR TITLE
VR-5552 Implement subclasses for _AutoscalingMetric

### PR DESCRIPTION
**Pull request by nhatsmrt**

### CODE REVIEW

**Comment by nhatsmrt** [comment]
@convoliution I have implemented all the metrics. I'm thinking of doing tests in https://github.com/VertaAI/modeldb/pull/1174.
For reference, the metrics are given by:
```
{
    "metrics": [
        {
            "parameters": [
                {
                    "type": "percentage", 
                    "description": "CPU utilization target", 
                    "name": "cpu_target"
                }
            ], 
            "description": "Trigger autoscaling based on CPU utilization", 
            "name": "cpu_utilization", 
            "metric_id": 1001
        }, 
        {
            "parameters": [
                {
                    "type": "count", 
                    "description": "Requests per worker target", 
                    "name": "requests_per_worker_target"
                }
            ], 
            "description": "Trigger autoscaling based on number of requests per worker", 
            "name": "requests_per_worker", 
            "metric_id": 1002
        }, 
        {
            "parameters": [
                {
                    "type": "percentage", 
                    "description": "Memory utilization target", 
                    "name": "memory_target"
                }
            ], 
            "description": "Trigger autoscaling based on memory utilization", 
            "name": "memory_utilization", 
            "metric_id": 1003
        }
    ]
}
```

File: client/verta/verta/deployment/autoscaling/metrics.py [hunk 1]
```diff
@@ -7,5 +7,63 @@
...
+    def _parameters_dict(self):
         raise NotImplementedError
+
+
+class CpuUtilization(_AutoscalingMetric):
+    _METRIC_ID = 1001
+    _NAME = "cpu_utilization"
+
+    def __init__(self, cpu_target):
+        self._cpu_target = cpu_target
```
Comment (convoliution) [inline client/verta/verta/deployment/autoscaling/metrics.py]: For all of these, could we also cast them to `str` for the user? I think it's reasonable for a user to expect to pass in a numerical value.

Plus, the fact that the backend wants `str`s is more of an implementation detail that the user shouldn't need to worry about (esp. if the backend's error message is that vague thing about Go type marshalling!)
```suggestion
        self._cpu_target = str(cpu_target)
```

**Review by convoliution — CHANGES_REQUESTED** [review]

File: client/verta/verta/deployment/autoscaling/metrics.py (hunk 1 above)
Comment (nhatsmrt) [reply client/verta/verta/deployment/autoscaling/metrics.py]: That makes perfect sense. Fixed.

**Review by convoliution — CHANGES_REQUESTED** [review]
Based on our discussion on Canary rules, can we make `_NAME` the name of the parameter rather than the parent metric?

I think this should be able to resemble [rules.py](https://github.com/VertaAI/modeldb/blob/master/client/verta/verta/deployment/update/rules.py) pretty closely, which puts all the logic in the parent class.

**Comment by nhatsmrt** [comment]
> Based on our discussion on Canary rules, can we make `_NAME` the name of the parameter rather than the parent metric?
> 
> I think this should be able to resemble [rules.py](https://github.com/VertaAI/modeldb/blob/master/client/verta/verta/deployment/update/rules.py) pretty closely, which puts all the logic in the parent class.

Done.

File: client/verta/verta/deployment/autoscaling/metrics.py [hunk 2]
```diff
@@ -7,5 +7,36 @@
 
 @six.add_metaclass(abc.ABCMeta)
 class _AutoscalingMetric(object):
-    def _as_json(self):
-        raise NotImplementedError
+    _METRIC_ID = 0
+    _NAME = ""
```
Comment (convoliution) [inline client/verta/verta/deployment/autoscaling/metrics.py]:
```suggestion
    _METRIC_ID = 0
    _PARENT_NAME = ""
    _NAME = ""
```

File: client/verta/tests/test_endpoint/test_endpoint.py [hunk 3]
```diff
@@ -318,9 +318,9 @@ def test_update_autoscaling(self, client, created_endpoints, experiment_run, mod
         created_endpoints.append(endpoint)
 
         autoscaling = Autoscaling(min_replicas=0, max_replicas=2, min_scale=0.5, max_scale=2.0)
-        autoscaling.add_metric(CpuUtilization("0.5"))
-        autoscaling.add_metric(MemoryUtilization("0.7"))
-        autoscaling.add_metric(RequestsPerWorker("100"))
+        autoscaling.add_metric(CpuTarget("0.5"))
+        autoscaling.add_metric(MemoryTarget("0.7"))
+        autoscaling.add_metric(RequestsPerWorkerTarget("100"))
```
Comment (convoliution) [inline client/verta/tests/test_endpoint/test_endpoint.py]: Could we support letting the user pass in numbers (at least for these metrics), and we convert it to a `str` as a courtesy? The fact that the backend API wants `str`s is an implementation detail; accepting numbers in numerical form seems nicer for the user. And the backend will let us know if the value doesn't make sense.

...Also I'm feeling some déjà vu here, but I genuinely can't remember. Was there somewhere else I brought up this idea?
```suggestion
        autoscaling.add_metric(CpuTarget(0.5))
        autoscaling.add_metric(MemoryTarget(0.7))
        autoscaling.add_metric(RequestsPerWorkerTarget(100))
```

File: client/verta/verta/deployment/autoscaling/metrics.py [hunk 4]
```diff
@@ -7,5 +7,36 @@
...
+
+    def __init__(self, value):
+        self._value = value
```
Comment (convoliution) [inline client/verta/verta/deployment/autoscaling/metrics.py]:
```suggestion
        self._value = str(value)
```

**Review by convoliution — CHANGES_REQUESTED** [review]
Let me know if these make sense

File: client/verta/verta/deployment/autoscaling/metrics.py (hunk 4 above)
Comment (nhatsmrt) [reply client/verta/verta/deployment/autoscaling/metrics.py]: Oh yeah the conversion was in the deleted code.

Comment (nhatsmrt) [reply client/verta/verta/deployment/autoscaling/metrics.py]: Fixed.

File: client/verta/tests/test_endpoint/test_endpoint.py (hunk 3 above)
Comment (nhatsmrt) [reply client/verta/tests/test_endpoint/test_endpoint.py]: Fixed.

File: client/verta/verta/deployment/autoscaling/metrics.py (hunk 2 above)
Comment (nhatsmrt) [reply client/verta/verta/deployment/autoscaling/metrics.py]: Fixed.

**Review by convoliution — APPROVED** [review]

**Pull request merged**